File: app/teryt_import.py
"""Importer plików XML z rejestru TERYT (GUS). Wspiera TERC, SIMC, ULIC."""

from __future__ import annotations

import xml.etree.ElementTree as ET
from typing import BinaryIO

from sqlalchemy import select
from sqlalchemy.orm import Session

from app import models


def _clean(val: str | None) -> str | None:
    if val is None: return None
    v = val.strip()
    return v if v else None


def import_terc_xml(db: Session, xml_file: BinaryIO) -> tuple[int, int]:
    """
    Importuje plik TERC (województwa, powiaty, gminy).
    """
    context = ET.iterparse(xml_file, events=("end",))
    s_count = 0
    d_count = 0
    
    states_data = {} # teryt_code -> name
    districts_data = [] # (woj_code, pow_code, name)
    
    for _, elem in context:
        if elem.tag == "row":
            woj = _clean(elem.findtext('WOJ'))
            powiat = _clean(elem.findtext('POW'))
            gmina = _clean(elem.findtext('GMI'))
            nazwa = _clean(elem.findtext('NAZWA'))
            
            if woj and not powiat and not gmina:
                states_data[woj] = nazwa.capitalize()
            elif woj and powiat and not gmina:
                districts_data.append((woj, powiat, nazwa))
            
            elem.clear()
            
    # Zapisz województwa
    state_map = {} # teryt_code -> models.LocationState.id
    for t_code, name in states_data.items():
        state = db.scalars(select(models.LocationState).where(models.LocationState.teryt_code == t_code)).first()
        if not state:
            state = models.LocationState(name=name, teryt_code=t_code)
            db.add(state)
            db.flush()
            s_count += 1
        state_map[t_code] = state.id
        
    # Zapisz powiaty
    for woj_code, pow_code, name in districts_data:
        state_id = state_map.get(woj_code)
        if not state_id:
            # Może województwo już było w bazie wcześniej?
            state = db.scalars(select(models.LocationState).where(models.LocationState.teryt_code == woj_code)).first()
            if state:
                state_id = state.id
                state_map[woj_code] = state_id
            else:
                continue
        
        exist = db.scalars(select(models.LocationDistrict).where(
            (models.LocationDistrict.state_id == state_id) & 
            (models.LocationDistrict.teryt_code == pow_code)
        )).first()
        if not exist:
            db.add(models.LocationDistrict(state_id=state_id, name=name, teryt_code=pow_code))
            d_count += 1
            
    db.commit()
    return s_count, d_count


def import_simc_xml(db: Session, xml_file: BinaryIO) -> int:
    """
    Importuje plik SIMC (miejscowości).
    """
    context = ET.iterparse(xml_file, events=("end",))
    count = 0
    
    dist_cache = {}
    all_districts = db.query(models.LocationDistrict).join(models.LocationState).all()
    for d in all_districts:
        dist_cache[(d.state.teryt_code, d.teryt_code)] = d.id

    for _, elem in context:
        if elem.tag == "row":
            woj = _clean(elem.findtext('WOJ'))
            powiat = _clean(elem.findtext('POW'))
            gmi = _clean(elem.findtext('GMI'))
            rodz = _clean(elem.findtext('RODZ_GMI')) or _clean(elem.findtext('RODZ'))
            nazwa = _clean(elem.findtext('NAZWA'))
            sym = _clean(elem.findtext('SYM'))
            
            if all([woj, powiat, nazwa, sym]):
                dist_id = dist_cache.get((woj, powiat))
                if dist_id:
                    exist = db.scalars(select(models.LocationCity).where(models.LocationCity.teryt_code == sym)).first()
                    if not exist:
                        db.add(models.LocationCity(
                            district_id=dist_id,
                            name=nazwa,
                            teryt_code=sym,
                            commune_code=gmi,
                            commune_type=rodz
                        ))
                        count += 1
                        if count % 500 == 0:
                            db.flush()
            elem.clear()
                
    db.commit()
    return count


def import_ulic_xml(db: Session, xml_file: BinaryIO) -> int:
    """
    Importuje plik ULIC (ulice). Wykorzystuje iterparse dla oszczędności pamięci.
    """
    context = ET.iterparse(xml_file, events=("end",))
    count = 0
    
    # Cache miast (SIMC)
    city_cache = {} # sym -> city_id
    all_cities = db.query(models.LocationCity.id, models.LocationCity.teryt_code).all()
    for cid, sym in all_cities:
        if sym: city_cache[sym] = cid

    for _, elem in context:
        if elem.tag == "row":
            sym = _clean(elem.findtext('SYM')) # SIMC miasta
            cecha = _clean(elem.findtext('CECHA')) # np. ul., al., pl.
            nazwa1 = _clean(elem.findtext('NAZWA_1'))
            nazwa2 = _clean(elem.findtext('NAZWA_2')) or ""
            ulic = _clean(elem.findtext('SYM_UL')) # Kod ulicy
            
            if all([sym, nazwa1, ulic]):
                city_id = city_cache.get(sym)
                if city_id:
                    full_name = f"{cecha} {nazwa2} {nazwa1}".strip().replace("  ", " ")
                    
                    # Sprawdzamy po ULIC w obrębie miasta
                    exist = db.scalars(select(models.LocationStreet).where(
                        (models.LocationStreet.city_id == city_id) & 
                        (models.LocationStreet.teryt_code == ulic)
                    )).first()
                    
                    if not exist:
                        db.add(models.LocationStreet(
                            city_id=city_id,
                            name=full_name,
                            teryt_code=ulic
                        ))
                        count += 1
                        if count % 1000 == 0:
                            db.flush()
            elem.clear()
                
    db.commit()
    return count

File: crm-portal-ts/src/teryt_import.ts
import { AppDataSource } from "./database.js";
import { LocationCity, LocationDistrict, LocationState, LocationStreet } from "./models/location.js";

const stateRepo = AppDataSource.getRepository(LocationState);
const districtRepo = AppDataSource.getRepository(LocationDistrict);
const cityRepo = AppDataSource.getRepository(LocationCity);
const streetRepo = AppDataSource.getRepository(LocationStreet);

function clean(value: string | null | undefined) {
    const trimmed = value?.trim() ?? "";
    return trimmed.length > 0 ? trimmed : null;
}

function extractRows(xmlContent: string) {
    return [...xmlContent.matchAll(/<row>([\s\S]*?)<\/row>/gi)].map((match) => match[1] ?? "");
}

function findTag(rowContent: string, tag: string) {
    const regex = new RegExp(`<${tag}>([\\s\\S]*?)<\\/${tag}>`, "i");
    const match = rowContent.match(regex);
    return clean(match?.[1]);
}

export async function importTercXml(xmlContent: string) {
    const rows = extractRows(xmlContent);
    const statesData = new Map<string, string>();
    const districtsData: Array<{ woj: string; powiat: string; name: string }> = [];

    for (const row of rows) {
        const woj = findTag(row, "WOJ");
        const powiat = findTag(row, "POW");
        const gmina = findTag(row, "GMI");
        const name = findTag(row, "NAZWA");

        if (woj && !powiat && !gmina && name) {
            statesData.set(woj, name);
        } else if (woj && powiat && !gmina && name) {
            districtsData.push({ woj, powiat, name });
        }
    }

    let importedStates = 0;
    let importedDistricts = 0;
    const stateIds = new Map<string, number>();

    for (const [terytCode, name] of statesData.entries()) {
        let state = await stateRepo.findOneBy({ terytCode });
        if (!state) {
            state = stateRepo.create({ name, terytCode });
            await stateRepo.save(state);
            importedStates += 1;
        }
        stateIds.set(terytCode, state.id);
    }

    for (const districtData of districtsData) {
        const stateId = stateIds.get(districtData.woj);
        if (!stateId) {
            continue;
        }

        const existingDistrict = await districtRepo.findOneBy({
            stateId,
            terytCode: districtData.powiat,
        });

        if (!existingDistrict) {
            const district = districtRepo.create({
                stateId,
                name: districtData.name,
                terytCode: districtData.powiat,
            });
            await districtRepo.save(district);
            importedDistricts += 1;
        }
    }

    return { importedStates, importedDistricts };
}

export async function importSimcXml(xmlContent: string) {
    const rows = extractRows(xmlContent);
    let importedCities = 0;

    const districts = await districtRepo.find({ relations: { state: true } });
    const districtMap = new Map<string, number>();
    for (const district of districts) {
        const stateCode = district.state?.terytCode;
        const districtCode = district.terytCode;
        if (stateCode && districtCode) {
            districtMap.set(`${stateCode}:${districtCode}`, district.id);
        }
    }

    for (const row of rows) {
        const woj = findTag(row, "WOJ");
        const powiat = findTag(row, "POW");
        const gmi = findTag(row, "GMI");
        const rodz = findTag(row, "RODZ_GMI") ?? findTag(row, "RODZ");
        const name = findTag(row, "NAZWA");
        const sym = findTag(row, "SYM");

        if (!woj || !powiat || !name || !sym) {
            continue;
        }

        const districtId = districtMap.get(`${woj}:${powiat}`);
        if (!districtId) {
            continue;
        }

        const existingCity = await cityRepo.findOneBy({ terytCode: sym });
        if (!existingCity) {
            const city = cityRepo.create({
                districtId,
                name,
                terytCode: sym,
                communeCode: gmi ?? undefined,
                communeType: rodz ?? undefined,
            });
            await cityRepo.save(city);
            importedCities += 1;
        }
    }

    return { importedCities };
}

export async function importUlicXml(xmlContent: string) {
    const rows = extractRows(xmlContent);
    let importedStreets = 0;

    const cities = await cityRepo.find();
    const cityMap = new Map<string, number>();
    for (const city of cities) {
        if (city.terytCode) {
            cityMap.set(city.terytCode, city.id);
        }
    }

    for (const row of rows) {
        const sym = findTag(row, "SYM");
        const cecha = findTag(row, "CECHA");
        const nazwa1 = findTag(row, "NAZWA_1");
        const nazwa2 = findTag(row, "NAZWA_2") ?? "";
        const ulic = findTag(row, "SYM_UL");

        if (!sym || !nazwa1 || !ulic) {
            continue;
        }

        const cityId = cityMap.get(sym);
        if (!cityId) {
            continue;
        }

        const name = `${cecha ?? ""} ${nazwa2} ${nazwa1}`.replace(/\s+/g, " ").trim();
        const existingStreet = await streetRepo.findOneBy({ cityId, terytCode: ulic });
        if (!existingStreet) {
            const street = streetRepo.create({
                cityId,
                name,
                terytCode: ulic,
            });
            await streetRepo.save(street);
            importedStreets += 1;
        }
    }

    return { importedStreets };
}
